### ISSUE TITLE
URI questions

### ISSUE THREAD

**Issue by atgreen** [joaotavora/snooze#15]
I was hoping to have URIs like...

http://localhost/api/v1/hello
http://localhost/api/v1/test

But I see that this doesn't work:
(defroute api/v1/hello ...

I don't see any mention of how method names map to paths.  How would I do something like this?

Thanks!

**Comment by atgreen** [comment]
When I do this...

(snooze:defroute api/v1/start (:get "text/plain")
  (setf *player-count* (+ 1 *player-count*))
  (format nil "~A" *player-count*))

I get a 404 error.  Hunchentoot reports this:

127.0.0.1 - [2018-11-20 08:38:39] "GET /api/v1/start/ HTTP/1.1" 404 13 "-" "Go-http-client/1.1"

**Comment by joaotavora** [comment]
It's not a bug, it's like this for design. Snooze provides a customization point for your case, `snooze:*resource-name-function*`. Here is its doc, which you can find in the api.lisp file:

```

**Comment by atgreen** [comment]
Yes, that works.

**Comment by atgreen** [comment]
Thanks.. I'll have a look later today!

**Comment by mdbergmann** [comment]
I'm also having some trouble here.
I didn't want to create a new ticket.

I hope I didn't overlook something but I've read the readme a few times now.

Basically I have two routes:

```
(defroute blog (:get :text/html)
(defroute blog (:get :text/html name)
```

When compiling this I get an error on the second definition:
```
Lambda list of method #<STANDARD-METHOD BLOG (SNOOZE-VERBS:GET
                                              SNOOZE-TYPES:TEXT/HTML)> 
is incompatible with that of the generic function BLOG.
Method's lambda-list : (SNOOZE-VERBS:HTTP-VERB SNOOZE-TYPES:TYPE)
Generic-function's   : (SNOOZE-VERBS:HTTP-VERB SNOOZE-TYPES:TYPE
                        NAME)
```

In principle it's clear, two methods with the same name. But my lacking CL knowledge doesn't let me overcome this.
Any ideas?

**Comment by joaotavora** [comment]
When you use `defroute` you are basically writing `defmethod`.  Just expand it (in `sly` it's `C-c C-m` on the `(` of the defroute.  In fact, you can `defmethod` directly.

Now, this should be enough to hint that a route in Snooze is only a method in a generic function.  So when you do the first `blog` you establish a certain protocol for the generic function, when you try to change that protocol, Lisp complains.  Most Lisps (I think SBCL at least) give you the option to replace the protocol of the generic function with  the new one.

You should learn about "Generic Functions".  In Snooze, a REST resource is represented by a generic function. The _methods_ of the generic functions are the different _routes_ that provide access to that resource.  Different types of access. Read access is a `GET` specializer on the first `HTTP-VERB` arg, write access if `PUT` or `POST` or `DELETE` on that arg.  And the same with the content types.

Also, in your case, it seems like you want a `/blog/<name-of-blog>` route. Ask yourself: is `<name-of-blog>` mandatory, or is it optional?  If it's optional it should be an `&optional` or `&key` argument.  If it's mandatory, then pointing your browser to just `/blog` automatically generates a 400 error, as one would expect.

**Comment by mdbergmann** [comment]
> or is it optional? If it's optional it should be an `&optional`

Great. This was a good pointer. Got it working with &optional.

> And then version and string will be bound to the uninterned symbols #v1 and #hello respectively, in the first case, and #v1 and #test in the second case. There are ample ways to configure this (see the README).

Are you refering to the `uri-to-arguments` method?

**Comment by mdbergmann** [comment]
I believe you can close this one as well.

**Comment by joaotavora** [comment]
> Are you refering to the uri-to-arguments method?

Yes, and its counterpart `arguments-to-uri`.    Whatever you do to `uri-to-arguments` you should _also_ do to `arguments-to-uri`, so that you can later use `blog-path` to generate perfectly matching URIs for your `blog` REST resource.

You could also use `read-for-resource` and `write-for-resource`, if you're concerned with the string representation of all the arguments for a particular resource.  Most times, you will want `uri-to-arguments` and `arguments-to-uri` though, leveraging their default implementations with `call-next-method`, of course.
